pkg/master/controller.go

```diff
@@ -143,31 +147,49 @@ func (c *Controller) CreateNamespaceIfNeeded(ns string) error {
 
 // createPortAndServiceSpec creates an array of service ports.
 // If the NodePort value is 0, just the servicePort is used, otherwise, a node port is exposed.
-func createPortAndServiceSpec(servicePort int, nodePort int) ([]api.ServicePort, api.ServiceType) {
+func createPortAndServiceSpec(servicePort int, nodePort int, servicePortName string, extraServicePorts []api.ServicePort) ([]api.ServicePort, api.ServiceType) {
 	//Use the Cluster IP type for the service port if NodePort isn't provided.
 	//Otherwise, we will be binding the master service to a NodePort.
-	if nodePort <= 0 {
-		return []api.ServicePort{{Protocol: api.ProtocolTCP,
+	servicePorts := []api.ServicePort{{Protocol: api.ProtocolTCP,
+		Port:       servicePort,
+		Name:       servicePortName,
+		TargetPort: util.NewIntOrStringFromInt(servicePort)}}
+	serviceType := api.ServiceTypeClusterIP
+	if nodePort > 0 {
+		servicePorts = []api.ServicePort{{Protocol: api.ProtocolTCP,
 			Port:       servicePort,
-			TargetPort: util.NewIntOrStringFromInt(servicePort)}}, api.ServiceTypeClusterIP
+			Name:       servicePortName,
+			TargetPort: util.NewIntOrStringFromInt(servicePort),
+			NodePort:   nodePort,
+		}}
+		serviceType = api.ServiceTypeNodePort
 	}
-	return []api.ServicePort{{Protocol: api.ProtocolTCP,
-		Port:       servicePort,
-		TargetPort: util.NewIntOrStringFromInt(servicePort),
-		NodePort:   nodePort,
-	}}, api.ServiceTypeNodePort
+	if extraServicePorts != nil {
+		servicePorts = append(servicePorts, extraServicePorts...)
```

Comment:
In light of the `nodePort` parameter, this reads weird.  It means that you'll start trying to claim additional ports on the nodes, right?  One for every `ExtraServicePort`.  However, you might not have set the `NodePort` field in the `extraServicePorts`.